src/MspFcComms.cpp

```diff
@@ -169,21 +172,125 @@ namespace FcComms
             return FcCommsReturns::kReturnError;
         }
 
+        // Try sending an ident request
+        MSP_IDENT ident;
+        sendMessage<MSP_IDENT>(ident);
+        char * const results = reinterpret_cast<char* const>(ident.response);
+        results[8] = '\0';
+        ROS_INFO("%s", ident.response);
         return FcCommsReturns::kReturnOk;
     }
 
+    // Implementation to send a receive a response from the flight controller
+    // Protocol specification here: http://www.stefanocottafavi.com/msp-the-multiwii-serial-protocol/
     template<typename T>
     FcCommsReturns MspFcComms::sendMessage(T& message)
     {
         if(fc_comms_status_ == FcCommsStatus::kConnected)
         {
-            // Send a message
-            // Check for errors
+            // Check length of data section
+            if(message.data_length > FcCommsMspConf::kMspMaxDataLength)
+            {
+                ROS_ERROR("FC_Comms data section > kMspMaxDataLength was attempted.");
+                return FcCommsReturns::kReturnError;
+            }
+
+            // Add the header, data_length, and message code
+            uint8_t packet[FcCommsMspConf::kMspNonDataLength + message.data_length];
+
+            std::copy(FcCommsMspConf::kMspSendHeader, FcCommsMspConf::kMspSendHeader + 3, packet);
+            packet[3] = message.data_length;
+            packet[4] = message.message_id;
+            
+            // Start off checksum calculation
+            uint8_t checksum{message.data_length ^ message.message_id};
+
+            #pragma GCC warning "Convert for loops in this function to some cleaner form of array copy"
+            // Copy data into message and finish calculating checksum
+            for(int i = 0; i < message.data_length; i++)
+            {
+                packet[FcCommsMspConf::kMspPacketDataOffset + i] = message.send[i];
+                checksum ^= message.send[i];
+            }
+
+            // Add checksum to packet
+            packet[FcCommsMspConf::kMspPacketDataOffset + message.data_length] = checksum;
+
+            try
+            {
+                fc_serial_->write(packet, FcCommsMspConf::kMspNonDataLength + message.data_length);
+            }
+            // Catch if there is an error writing
+            catch(const std::exception& e)
+            {
+                ROS_ERROR("FC_Comms error sending MSP packet");
+                ROS_ERROR("Exception: %s", e.what());
+                return FcCommsReturns::kReturnError;
+            }
+
+            #pragma GCC warning "It would be good to split this off to another function"
+            // Now receive
+            if(message.has_response)
+            {
+                std::string header = fc_serial_->read(FcCommsMspConf::kMspHeaderSize);
+                if(header == FcCommsMspConf::kMspReceiveHeader)
```

Comment:
You're right the logic is incorrect. We can do the comparison this way because header is a c++ string obje ct and the comparison operator is overloaded for comparing a string object and a char array. I will be more explicit about it though. http://www.cplusplus.com/reference/string/string/operators/

I believe strncmp is more of a C style thing and C++ strings are preferred.